### ISSUE TITLE
Keycloak group not synced when logged into sonar

### ISSUE THREAD

**Issue by rajith77** [vaulttec/sonar-auth-oidc#60]
Sonar version 9.2
Plugin version 2.1.1

Keycloak is used as the OIDC provider
We have a user with a group and have added the group mapper.
We have a corresponding group in sonarqube too.

However when the user is logged into Sonar his group is not synced and the user is placed in the sonar-users group.
With debug logging enabled we don't see any logs related to OIDC.
But I do see the following log repeatedly printed
login failure [cause|User must be authenticated][method|BASIC][provider|LOCAL|local][IP|10.128.0.2|10.118.48.4][login|]

Even though that log is printed the user is able to log into sonarqube.
But I feel the above error indicates that the authentication process doesn't complete properly hence the groups are not fetched.

**Comment by tjuerge** [comment]
> With debug logging enabled we don't see any logs related to OIDC.

Is the plugin [installed, enabled and configured](https://github.com/vaulttec/sonar-auth-oidc#installation) properly? Is the [plugins login button](https://github.com/vaulttec/sonar-auth-oidc#description) visible in SQ?

**Comment by tjuerge** [comment]
> Same problem! I had checked whether the plugin was installed, enabled and configured. And the plugins login button is visible. I also created a mapper in keycloak idp, but still made no sense.

Please check [SonarQube's debug log](https://github.com/vaulttec/sonar-auth-oidc#troubleshooting).

> In fact, I have another question that the plugin can not associate with user existed in SonaQube.

Starting with SQ 8 (due to [SONAR-12475](https://jira.sonarsource.com/browse/SONAR-12475)) it's [not possible anymore to create users in SQ with specific login name to associate it with an existing user](https://community.sonarsource.com/t/how-to-pre-create-users-who-will-authenticate-via-an-identity-provider-with-sq-8/24207). More details can be found in the comments to #41 or #42.

**Comment by paigeman** [comment]
> Please check [SonarQube's debug log](https://github.com/vaulttec/sonar-auth-oidc#troubleshooting).

I checked the SQ's debug log and found records below.
```log
2022.08.24 00:56:29 DEBUG web[AYLNWLF9BTuMz4k1AAAc][o.v.s.a.o.OidcClient] User info: {"at_hash":"ezSVbyWqV5xu51v9bytStA","sub":"f1277534-78c8-4664-af06-547d62afce9a","email_verified":false,"iss":"","typ":"ID","preferred_username":"","given_name":"fade","sid":"e31994a1-50fa-4ff6-bafc-255a38b67487","aud":"SonarQube-oidc","acr":"1","azp":"SonarQube-oidc","auth_time":1661302589,"name":"","exp":1661302889,"session_state":"e31994a1-50fa-4ff6-bafc-255a38b67487","iat":1661302589,"family_name":"","jti":"53c8cb6b-c497-419f-b3f3-2b57a8ba480a","email":"","group":"sonar-administrators"}
2022.08.24 00:56:29 DEBUG web[AYLNWLF9BTuMz4k1AAAc][o.v.s.a.o.OidcIdentityProvider] Authenticating user '' with groups []
```
As you can see, you can see my custom claim **group** at the end of the first line.
However, the second line shows that it authenticated user with groups.
Here is the screenshot of my SQ's configuration:
![图片](https://user-images.githubusercontent.com/53284808/186295667-a5eb9354-da85-4b79-a2a8-606fa7efb061.png)
Any useful tips?

**Comment by tjuerge** [comment]
> However, the second line shows that it authenticated user with groups.

It shows that no user login was generated by the plugin. Which login generation strategy did you select in the plugin's settings? Depending on the selected login generation strategy your IdP must provide the required data in the UserInfo, e.g. `preferred_username` or `email`.

**Comment by paigeman** [comment]
> > However, the second line shows that it authenticated user with groups.
> 
> It shows that no user login was generated by the plugin (empty single quotes). Which login generation strategy did you select in the plugin's settings? Depending on the selected login generation strategy your IdP must provide the required data in the UserInfo, e.g. `preferred_username` or `email`.

sorry for disturbing you, I had deleted sensitive info when I commented. I can actually log in SQ, just can not use custom group claim

**Comment by tjuerge** [comment]
The groups claim is expected to be marshalled as a JSON string array (explained in #42). This is also valid for a single group.

**Comment by tjuerge** [comment]
> > I can actually log in SQ, just can not use custom group claim
> 
> The groups claim is expected to be marshalled as a JSON string array (explained in #42). This is also valid for a single group.

Support for groups claim marshalled as string will be implented in #64.

**Comment by paigeman** [comment]
> > > I can actually log in SQ, just can not use custom group claim
> > 
> > 
> > The groups claim is expected to be marshalled as a JSON string array (explained in #42). This is also valid for a single group.
> 
> Support for groups claim marshalled as a JSON string (instead of a JSON string array) is implemented in #64.

Thanks, I have tested it, it works

**Comment by tjuerge** [comment]
> Thanks, I have tested it, it works

Cool, then we can close this ticket as well 👍
